diff --git a/core/protorpc/protoc-gen-protorpc/main.go b/core/protorpc/protoc-gen-protorpc/main.go
index ab15ff4..34b5015 100644
--- a/core/protorpc/protoc-gen-protorpc/main.go
+++ b/core/protorpc/protoc-gen-protorpc/main.go
@@ -42,7 +42,7 @@ func (p *protorpcPlugin) HeaderCode(g *generator.Generator, file *generator.File
 //
 // source: {{$File.GetName}}
 
-package {{$File.PackageName}}
+package gen
 
 import (
 	"fmt"
