--- a/source/core/slang-stream.cpp
+++ b/source/core/slang-stream.cpp
@@ -204,15 +204,11 @@ Int64 FileStream::getPosition()
 {
 #if defined(_WIN32) || defined(__CYGWIN__)
     fpos_t pos;
     fgetpos(m_handle, &pos);
     return pos;
-#elif defined(__APPLE__)
-    return ftell(m_handle);
 #else
-    fpos64_t pos;
-    fgetpos64(m_handle, &pos);
-    return *(Int64*)(&pos);
+    return ftello(m_handle);
 #endif
 }
 
 SlangResult FileStream::seek(SeekOrigin seekOrigin, Int64 offset)
